### PR TITLE
Rematch5 compatibility 

### PR DESCRIPTION
**Pull request by bloerwald**

### CODE REVIEW

File: Rematch/UI.lua [hunk 1]
```diff
@@ -9,251 +9,93 @@ local ns = select(2, ...)
...
-            self:RemoveScript(key)
-            return origAccept(...)
+        if team and script then
+            script:SetName(team.name)
         end
-    end, true)
+    end)
 
-    -- team rename
+    -- TODO: what do we want to do when a team is overwritten? delete the script? keep it? migrate the old script to the new teamID?
```
Comment (bloerwald) [inline Rematch/UI.lua:46]: open todo

>  what do we want to do when a team is overwritten? delete the script? keep it? migrate the old script to the new teamID?

File: Rematch/UI.lua [hunk 2]
```diff
@@ -9,251 +9,93 @@ local ns = select(2, ...)
...
-        CollectionsJournal:SetAttribute('UIPanelLayout-width', 710)
-        UpdateUIPanelPositions()
-    end)
 end
 
+
 function Addon:OnDisable()
-    tDeleteItem(Rematch:GetMenu('TeamMenu'), teamMenu)
-end
+    -- TODO: hide menu, disable badge?
```
Comment (bloerwald) [inline Rematch/UI.lua:78]: open todo

> hide menu, disable badge?

File: Rematch/UpdateDB.lua [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    end
+
+    -- now we can migrade scripts.
+    for key, script in pairs(scriptList) do
+        local newTeamID = convertedTeams[key]
+        if newTeamID then
+            self:RemoveScript(key)
+            self:AddScript(newTeamID, script)
+
+            --TODO: do we want to maintain the same script name and team name?
```
Comment (bloerwald) [inline Rematch/UpdateDB.lua:30]: open todo

> do we want to maintain the same script name and team name?

File: Rematch/UI.lua [hunk 4]
```diff
@@ -9,251 +9,93 @@ local ns = select(2, ...)
...
-    end
+    --if Rematch.savedTeams[key] then
+    --    return Rematch.teamStrings:ExportTeam(key)
+    --end
 end
 
 function Addon:OnImport(data)
-    Rematch:ShowImportDialog()
-    RematchDialog.MultiLine.EditBox:SetText(data)
+    -- TODO: Bug. (Do people even use this?)
```
Comment (bloerwald) [inline Rematch/UI.lua:95]: open todo

> TODO: Bug. (Do people even use this?)
> - When importing a PBS script, the key is preset in the data, but it wont match up after importing this Rematch team.

File: Rematch/UI.lua [hunk 5]
```diff
@@ -9,251 +9,93 @@ local ns = select(2, ...)
...
+    -- be made orphans, and if a new team is created with that same key, the
+    -- old script will be matched with it.
 end
 
 function Addon:OnExport(key)
-    if RematchSaved[key] then
-        Rematch:SetSideline(key, RematchSaved[key])
-        return Rematch:ConvertSidelineToString()
-    end
+    --if Rematch.savedTeams[key] then
```
Comment (bloerwald) [inline Rematch/UI.lua:89]: open todo: export not implemented

File: Rematch/Addon.lua [hunk 6]
```diff
@@ -18,33 +18,33 @@ function Addon:OnInitialize()
 end
 
 function Addon:GetCurrentKey()
-    return RematchSettings.loadedTeam
+    return Rematch.settings.currentTeamID
```
Comment (bloerwald) [inline Rematch/Addon.lua:21]: open question/todo: maintain rematch 4 compatibility for the transition period?

File: Rematch/UI.lua (hunk 1 above)
Comment (gizzmo) [reply Rematch/UI.lua:46]: I don't fully understand, from the callback arguments alone, how to determine what exactly is happening.

File: Rematch/UI.lua (hunk 2 above)
Comment (gizzmo) [reply Rematch/UI.lua:78]: * I don't see an easy way to remove menu entries. 
* Badges are easy to disable with a check in the callback.

Is it best to just reload after this is disabled?

File: Rematch/UpdateDB.lua (hunk 3 above)
Comment (gizzmo) [reply Rematch/UpdateDB.lua:30]: In `UpdateDB.lua` I'm already doing that on migration. I think doing it here would be ok.

File: Rematch/Addon.lua (hunk 6 above)
Comment (gizzmo) [reply Rematch/Addon.lua:21]: Making use of that `Version.lua` file/class to load the hooks/events depending on the version of rematch is the way to go there. 

But the real issue is loading the scripts from the database. The key in the database key is the profile name, and it appears the script's for that plugin are loaded before the plugin is even enabled. 

Maybe we can use a separate plugin for Rematch5, and check which version of Rematch has loaded when the both plugins are enabled?

**Comment by axc450** [comment]
Thanks all for all work on this. This is a big milestone to get working with Rematch 5. I think supporting Rematch 4 for now is OK (but it should be easy to remove in the future)

**Comment by gizzmo** [comment]
> Thanks all for all work on this.

My pleasure, I really wanted to try the Rematch 5 beta, but can't live without this amazing addon. (Bonus I enjoy coding 😉)

**Comment by bloerwald** [comment]
> The initial load needs to happen either at the same time, or after Rematch5.
> 
> If you've used Rematch5, do a /rematch reset everything
> - Logout and apply my changes
> - Log into the game and pray things work.

We can't keep that UX, people will complain too much since our releases are unlikely to line up perfectly.

**Comment by bloerwald** [comment]
Bug: Script manager has all script names, but shows neither the matching team, nor the script content.

**Comment by bloerwald** [comment]
UX: Notes in teams list is a button but we only get a badge. People will want the old UX back where clicking the badge opens the editor. Check with Rematch whether we can have that again.

File: Rematch/UpdateDB.lua (hunk 3 above)
Comment (gizzmo) [reply Rematch/UpdateDB.lua:30]: Oh whoops this comment is in the `UpdateDB.lua` file. That sync is already happening in the `REMATCH_TEAM_UPDATED` event, so i don't see why it shouldn't be here too.

File: Rematch/Addon.lua (hunk 6 above)
Comment (bloerwald) [reply Rematch/Addon.lua:21]: Once `RematchPlugin:OnEnabled()` is called, rematch is loaded. The bad part is that at some times `RematchPlugin:*` is called without `:OnEnabled()` ever being called. I think I got all those cases covered, making it now possible to use version everywhere with local conditions and thus little duplication.

File: Rematch/UI.lua (hunk 4 above)
Comment (bloerwald) [reply Rematch/UI.lua:95]: I don't actually think people use this, but I kind of would love if people do. One of the more common requests on xufu is importing team and script at the same time, which would work with this. Thus, I'd prefer not to remove it.

File: Rematch/UI.lua (hunk 2 above)
Comment (bloerwald) [reply Rematch/UI.lua:78]: - done for rematch 4: menu, badges, events
- done for rematch 5: events
- missing for rematch 5: menu, badges

File: Rematch/UI.lua (hunk 1 above)
Comment (bloerwald) [reply Rematch/UI.lua:46]: asked gello in https://www.warcraftpets.com/community/forum/viewtopic.php?p=149579, but it *seems* right

File: Rematch/UI.lua (hunk 2 above)
Comment (bloerwald) [reply Rematch/UI.lua:78]: rematch5 menu and badges requires new api. there are two working hacks around it in #61, so done as well

File: Rematch/UI.lua (hunk 5 above)
Comment (bloerwald) [reply Rematch/UI.lua:89]: export and import now works

**Comment by bloerwald** [comment]
closing: #61 was superset and is merged

**Pull request closed without merging**